src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java

```diff
@@ -2849,26 +2842,55 @@ public void watchFiles(File outputDirectory, File testOutputDirectory, final Thr
-                            if (isMultiModuleProject()) {
-                                runTestThread(false, executor, -1, false, getAllBuildFiles());
-                            } else {
-                                runTestThread(false, executor, -1, false, false, buildFile);
+                // Assumption is that initialCompile will only be set to false after ALL modules are complete
+                if (!recompileDependencies && !registerClassDirAttempted && !initialCompile) {
+                    debug("Attempt to register class file directories for feature generation");
+                    // set to true regardless of success
+                    // this may need to be updated, but currently do not want to come into this section again
+                    registerClassDirAttempted = true;
+                    if (isMultiModuleProject()) {
+                        for (ProjectModule p : upstreamProjects) {
+                            if (shouldIncludeSources(p.getPackagingType())) {
+                                // register source classes directory for other modules
+                                if (p.getOutputDirectory().exists()) {
+                                    registerAll(p.getOutputDirectory().getCanonicalFile().toPath(), executor);
+                                    p.sourceOutputDirRegistered = true;
+                                }
                             }
                         }
                     }
+                    if (shouldIncludeSources(packagingType)) {
+                         // register source classes directory for main module
+                        if (this.outputDirectory.exists()) {
+                            registerAll(srcOutputPath, executor);
+                            sourceOutputDirRegistered = true;
+                        }
+                    }
+                }
+
+                // Generate features from source file changes
+                if (generateFeatures && !classesFailingToCompile()) { //do not run generate features if there are classes failing to compile
+                    boolean tryFeatureGeneration = false;
+                    if (recompileDependencies && lastChangeCompiled && !javaSourceClasses.isEmpty()) { // no class file tracking
+                        debug("Detected a change in the following class directories: " + javaSourceClasses);
+                        lastChangeCompiled = false;
+                        modifiedSrcBuildFile = null;
+                        tryFeatureGeneration = true;
+                        incrementGenerateFeatures();
+                    } else if (!recompileDependencies && !javaSourceClasses.isEmpty()) { // class file tracking
+                        debug("Detected a change in the following classes: " + javaSourceClasses);
+                        tryFeatureGeneration = true;
+                        incrementGenerateFeatures();
+                    }
```

Comment:
I think this could be simplified to remove the `tryFeatureGeneration` boolean (just a suggestion): 
```java
if (generateFeatures && !classesFailingToCompile && !javaSourceClasses.isEmpty()  {
    if ((recompileDependencies && lastChangeCompiled) || !recompileDependencies) {
        if (recompileDependencies) {
            // no class file tracking
            debug("Detected a change in the following class directories: " + javaSourceClasses);
            lastChangeCompiled = false;
            modifiedSrcBuildFile = null;
        } else {
             // class file tracking
            debug("Detected a change in the following classes: " + javaSourceClasses);
        }
        incrementGenerateFeatures();
        if (!generatedFeaturesFile.exists()) {
             // run tests logic
        }
    }
}
```